zvariant/src/fd.rs

```diff
@@ -1,33 +1,109 @@
 use serde::{de::Error, Deserialize, Deserializer, Serialize, Serializer};
 use static_assertions::assert_impl_all;
-use std::os::{fd::IntoRawFd, unix::io};
+use std::os::fd::{self, AsFd, AsRawFd, BorrowedFd, RawFd};
 
-use crate::{Basic, EncodingFormat, Signature, Type};
+use crate::{serialized::Format, Basic, Signature, Type};
 
-/// A [`RawFd`](https://doc.rust-lang.org/std/os/unix/io/type.RawFd.html) wrapper.
+/// A file-descriptor type wrapper.
 ///
-/// See also `OwnedFd` if you need a wrapper that takes ownership of the file.
-///
-/// We wrap the `RawFd` type so that we can implement [`Serialize`] and [`Deserialize`] for it.
-/// File descriptors are serialized in a special way and you need to use specific [serializer] and
-/// [deserializer] API when file descriptors are or could be involved.
+/// Since [`std::os::fd::BorrowedFd`] and [`std::os::fd::OwnedFd`] types
+/// do not implement  [`Serialize`] and [`Deserialize`]. So we provide a
+/// wrapper for both that implements these traits.
 ///
 /// [`Serialize`]: https://docs.serde.rs/serde/trait.Serialize.html
 /// [`Deserialize`]: https://docs.serde.rs/serde/de/trait.Deserialize.html
-/// [deserializer]: fn.from_slice_fds.html
-/// [serializer]: fn.to_bytes_fds.html
-#[derive(Debug, PartialEq, PartialOrd, Ord, Eq, Hash, Clone, Copy)]
-pub struct Fd(io::RawFd);
+#[derive(Debug)]
+pub enum Fd<'f> {
+    Borrowed(BorrowedFd<'f>),
+    Owned(fd::OwnedFd),
+}
+
+impl<'f> Fd<'f> {
+    /// Try to create an owned version of `self`.
+    pub fn try_to_owned(&self) -> crate::Result<Fd<'static>> {
+        self.as_fd()
+            .try_clone_to_owned()
+            .map(Fd::Owned)
+            .map_err(Into::into)
+    }
+
+    /// Try to clone `self`.
+    pub fn try_clone(&self) -> crate::Result<Self> {
+        Ok(match self {
+            Self::Borrowed(fd) => Self::Borrowed(*fd),
+            Self::Owned(fd) => Self::Owned(fd.try_clone()?),
+        })
+    }
+}
+
+impl<'f> From<BorrowedFd<'f>> for Fd<'f> {
+    fn from(fd: BorrowedFd<'f>) -> Self {
+        Self::Borrowed(fd)
+    }
+}
+
+impl From<fd::OwnedFd> for Fd<'static> {
+    fn from(fd: fd::OwnedFd) -> Self {
+        Self::Owned(fd)
+    }
+}
+
+impl From<OwnedFd> for Fd<'static> {
+    fn from(owned: OwnedFd) -> Self {
+        owned.inner
+    }
+}
+
+impl TryFrom<Fd<'static>> for fd::OwnedFd {
```

Comment:
Now that we clone, could this requirement be soften up to a random lifetime `'fd`?. Besides that, MR looks :ok_hand:  to me.